phylo/src/alignment/alignment_builder.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+
+use anyhow::bail;
+
+use crate::align;
+use crate::alignment::{Alignment, InternalMapping, Mapping, PairwiseAlignment, Sequences};
+use crate::tree::{NodeIdx, NodeIdx::Internal as Int, NodeIdx::Leaf, Tree};
+use crate::Result;
+
+pub struct AlignmentBuilder<'a> {
+    tree: &'a Tree,
+    seqs: Sequences,
+    node_map: InternalMapping,
+}
+
+impl<'a> AlignmentBuilder<'a> {
+    pub fn new(tree: &'a Tree, seqs: Sequences) -> AlignmentBuilder<'a> {
+        AlignmentBuilder {
+            tree,
+            seqs,
+            node_map: InternalMapping::new(),
+        }
+    }
+
+    pub fn msa(mut self, msa: InternalMapping) -> Self {
+        self.node_map = msa;
+        self
+    }
+
+    pub fn build(self) -> Result<Alignment> {
+        if self.node_map.is_empty() {
```

Comment:
nit: I would recommend reversing the condition here, as it would reduce the levels in the function:
```rust
if !self.node_map.is_empty() {
  self.build_from_map()
} else if self.seqs.aligned {
  self.build_from_seqs()
} else {
  self.build_from_unaligned()
}
```

Reply:
Yes, makes perfect sense (;